### PR TITLE
Suppress Write-Progress in ConsoleHost if output is redirected and fix tests

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

When output is redirected, Write-Progress rendering of the ConsoleHost tries to hide the cursor (or show the cursor) which throws an exception.  Fix is to not render any progress at all if output is redirected as it's likely used for automation anyways.

The Disconnect test seems to be failing where it validates the session is in a `Disconnected` state, but typically has been in a `Disconnecting` state.  Updated test to handle both `Disconnected` and `Disconnecting` to allow it to pass.

Markdown URL link tests sometimes fail when the 3rd party site timesout.  Added a status code to see what is being returned and whether we should allow that to pass.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1 [hunk 1]
```diff
@@ -473,8 +473,8 @@ Describe "Additional tests for Import-Module with WinCompat" -Tag "Feature" {
 
         It "Verify that Warning is generated with default WarningAction" {
             $LogPath = Join-Path $TestDrive (New-Guid).ToString()
-            & $pwsh -NoProfile -NonInteractive -c "[System.Management.Automation.Internal.InternalTestHooks]::SetTestHook('TestWindowsPowerShellPSHomeLocation', `'$basePath`');Import-Module $ModuleName" *> $LogPath
-            $LogPath | Should -FileContentMatch 'loaded in Windows PowerShell'
+            & $pwsh -NoProfile -NonInteractive -c "[System.Management.Automation.Internal.InternalTestHooks]::SetTestHook('TestWindowsPowerShellPSHomeLocation', `'$basePath`');Import-Module $ModuleName;Get-Error" *> $LogPath
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1:476]: Why `Get-Error` is needed now?

Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1:476]: Needed this to figure out why this was failing.  It only shows up when the test fails.  Needed the stacktrace.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ProgressPane.cs [hunk 2]
```diff
@@ -301,7 +301,15 @@ private void WriteContent()
...
-                Console.CursorVisible = false;
+                try
+                {
+                    Console.CursorVisible = false;
+                }
+                catch (System.IO.IOException)
+                {
+                    // ignore if set can't set cursor visibility
+                }
+
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ProgressPane.cs:312]: I'm able to reproduce the issue by redirecting the standard out of the pwsh process:

image

Maybe the right fix would be to not render the progress when the standard output of the current process is redirected? Or revert back to the `Classic` view when stdout is redirected? (I personally think the former makes more sense).

File: test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1 (hunk 1 above)
Comment (daxian-dbw) [reply test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1:476]: So, I guess if the test run successfully, `Get-Error` shouldn't return anything, which means only the expected warning message is written to the stdout stream?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ProgressPane.cs (hunk 2 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ProgressPane.cs:312]: Yeah, that would make sense to not render progress at all.

Comment (daxian-dbw) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ProgressPane.cs:305]: I refactored the `WriteProgress` method body a little bit to reduce nesting there. It's easier to see the difference by ignoring the whitespaces: https://github.com/PowerShell/PowerShell/pull/14716/files?w=1

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ProgressPane.cs:305]: LGTM

File: test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1 (hunk 1 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1:476]: yes

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**